Mage.Sets/src/mage/cards/r/RiverSongsDiary.java

```diff
@@ -0,0 +1,195 @@
+package mage.cards.r;
+
+import mage.MageObjectReference;
+import mage.abilities.Ability;
+import mage.abilities.condition.Condition;
+import mage.abilities.decorator.ConditionalInterveningIfTriggeredAbility;
+import mage.abilities.effects.Effect;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.ReplacementEffectImpl;
+import mage.abilities.TriggeredAbilityImpl;
+import mage.abilities.triggers.BeginningOfUpkeepTriggeredAbility;
+import mage.cards.Card;
+import mage.cards.Cards;
+import mage.cards.CardImpl;
+import mage.cards.CardsImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.*;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.events.ZoneChangeEvent;
+import mage.game.ExileZone;
+import mage.game.stack.Spell;
+import mage.players.Player;
+import mage.target.targetpointer.FixedTarget;
+import mage.util.CardUtil;
+
+import java.util.UUID;
+
+/**
+ * @author padfoothelix 
+ */
+public final class RiverSongsDiary extends CardImpl {
+
+    public RiverSongsDiary(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ARTIFACT}, "{3}");
+        
+
+        // Imprint -- Whenever a player casts an instant or sorcery spell from their hand, exile it instead of putting it into a graveyard as it resolves.
+	this.addAbility(new RiverSongsDiaryImprintAbility().setAbilityWord(AbilityWord.IMPRINT));
+
+	// At the beginning of your upkeep, if there are four or more cards exiled with River Song's Diary, choose one of them at random. You may cast it without paying its mana cost.
+	this.addAbility(new ConditionalInterveningIfTriggeredAbility(
+	        new BeginningOfUpkeepTriggeredAbility(
+		        new RiverSongsDiaryCastEffect()
+		), RiverSongsDiaryCondition.instance, 
```

Comment:
I recommend to add card hints to any abilities with condition usage. For your use case -- value card hint with exiled cards amount. See usage of `ValueConditionHint`.